AX: Auto-expand details and hidden="until-found" elements containing search text provided in a AXUIElementsForSearchPredicate request

twilco opened the pull request:
#### bec672f7d9f7fbd72f27dbf598ae69badafcaa75
<pre>
AX: Auto-expand details and hidden=&quot;until-found&quot; elements containing search text provided in a AXUIElementsForSearchPredicate request
<a href="https://bugs.webkit.org/show_bug.cgi?id=298432">https://bugs.webkit.org/show_bug.cgi?id=298432</a>
<a href="https://rdar.apple.com/159913471">rdar://159913471</a>

Reviewed by Joshua Hoffman.

If the user is searching for a specific string, as indicated by search text being present in a AXUIElementsForSearchPredicate
request, we should provide matches if said string is inside a collapsed details element or hidden=&quot;until-found&quot; container,
making sure to expand them as well.

This gets a bit tricky in the case of the accessibility thread, which obviously cannot inherently synchronously expand
an element. This patch adds a new `performFunctionOnMainThreadAndWaitWithTimeout` function to attempt allow
the accessibility thread to attempt a synchronous action, but bail after a timeout if the main-thread is busy. We use
this to try to expand revealable elements synchronously, but if we can&apos;t, we&apos;ll continue on with the search.

If we do timeout trying to reveal a container, we won&apos;t perform any further synchronous trips to the main-thread for
the remainder of the search, with the assumption being that the main-thread is busy doing other things.

New property AXProperty::RevealableText is added and cached for text objects within an auto-revealable container. Another
property, AXProperty::IsHiddenUntilFoundContainer, is also added to indicate that we should search within the given object
for revealable text.

* LayoutTests/accessibility/mac/text-search-auto-expands-revealable-elements-expected.txt: Added.
* LayoutTests/accessibility/mac/text-search-auto-expands-revealable-elements.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::revealableContainers):
(WebCore::AXCoreObject::detailsAncestor const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::Accessibility::performFunctionOnMainThreadAndWaitWithTimeout):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXNotifications.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::documentNeedsLayoutOrStyleRecalc): Deleted.
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::AXSearchManager::revealHiddenMatchWithTimeout):
(WebCore::AXSearchManager::findMatchingObjectsInternal):
* Source/WebCore/accessibility/AXSearchManager.h:
(WebCore::AXSearchManager::lastRevealAttemptTimedOut):
(WebCore::AXSearchManager::setLastRevealAttemptTimedOut):
* Source/WebCore/accessibility/AXUtilities.cpp:
(WebCore::needsLayoutOrStyleRecalc):
* Source/WebCore/accessibility/AXUtilities.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::revealAncestors):
(WebCore::AccessibilityNodeObject::isHiddenUntilFoundContainer const):
(WebCore::AccessibilityNodeObject::revealableText const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::defaultObjectInclusion const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::revealAncestors):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::convertToPropertyFlag):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::objectBecameIgnored):

Canonical link: <a href="https://commits.webkit.org/299649@main">https://commits.webkit.org/299649@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d1e994ee0cf84ade5306aaf2dfcc3c704445539e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119687 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39379 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30031 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125980 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71759 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ce17d6c1-7e39-4db5-b2aa-5edd10770b71) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121563 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40075 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47957 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90908 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/60201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c66d7380-b0b4-47e3-a5bf-4f3af5fef398) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122639 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31994 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107319 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71435 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e17e9e07-3e42-4832-887a-ba3827b548ac) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31028 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25423 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69619 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101462 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25615 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128933 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46607 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35308 "Found 1 new test failure: http/tests/cache/disk-cache/disk-cache-302-status-code.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99504 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/forms/constraints/infinite_backtracking.html (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46972 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103511 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99347 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25234 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44777 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22798 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43172 "Hash d1e994ee for PR 50347 does not build (failure)") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46469 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52175 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45935 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49284 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47621 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -3318,6 +3334,20 @@ String AccessibilityNodeObject::textUnderElement(TextUnderElementMode mode) cons
+        return nullString();
+
+    CheckedPtr<const RenderStyle> style = this->style();
+    if (!style || !style->autoRevealsWhenFound())
```
If something is already "found", is `autoRevealsWhenFound` false? Trying to reconcile when `revealableText` will return an empty string

twilco replied:
Correct, yeah if a details is expanded for example, the text within will be `!autoRevealsWhenFound()`

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3811,8 +3811,15 @@ AccessibilityObjectInclusion AccessibilityObject::defaultObjectInclusion() const
+                // We don't want to ignore hidden-until-found containers because we
+                // want AXSearchManager::findMatchingObjects to walk over them in order
+                // to search within them for revealable text matching the search query string.
+                return AccessibilityObjectInclusion::IncludeObject;
```
Could this ever lead to things that should be ignored to not be? E.g., what if you have a container that for some reason is hidden-until-found, but is also aria-hidden?

twilco replied:
Great point! Fixed in latest commit and new testcase added ensuring we respect aria-hidden.

twilco committed:
AX: Auto-expand details and hidden="until-found" elements containing search text provided in a AXUIElementsForSearchPredicate request
https://bugs.webkit.org/show_bug.cgi?id=298432
rdar://159913471

Reviewed by Joshua Hoffman.

If the user is searching for a specific string, as indicated by search text being present in a AXUIElementsForSearchPredicate
request, we should provide matches if said string is inside a collapsed details element or hidden="until-found" container,
making sure to expand them as well.

This gets a bit tricky in the case of the accessibility thread, which obviously cannot inherently synchronously expand
an element. This patch adds a new `performFunctionOnMainThreadAndWaitWithTimeout` function to attempt allow
the accessibility thread to attempt a synchronous action, but bail after a timeout if the main-thread is busy. We use
this to try to expand revealable elements synchronously, but if we can't, we'll continue on with the search.

If we do timeout trying to reveal a container, we won't perform any further synchronous trips to the main-thread for
the remainder of the search, with the assumption being that the main-thread is busy doing other things.

New property AXProperty::RevealableText is added and cached for text objects within an auto-revealable container. Another
property, AXProperty::IsHiddenUntilFoundContainer, is also added to indicate that we should search within the given object
for revealable text.

* LayoutTests/accessibility/mac/text-search-auto-expands-revealable-elements-expected.txt: Added.
* LayoutTests/accessibility/mac/text-search-auto-expands-revealable-elements.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::revealableContainers):
(WebCore::AXCoreObject::detailsAncestor const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::Accessibility::performFunctionOnMainThreadAndWaitWithTimeout):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXNotifications.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::documentNeedsLayoutOrStyleRecalc): Deleted.
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::AXSearchManager::revealHiddenMatchWithTimeout):
(WebCore::AXSearchManager::findMatchingObjectsInternal):
* Source/WebCore/accessibility/AXSearchManager.h:
(WebCore::AXSearchManager::lastRevealAttemptTimedOut):
(WebCore::AXSearchManager::setLastRevealAttemptTimedOut):
* Source/WebCore/accessibility/AXUtilities.cpp:
(WebCore::needsLayoutOrStyleRecalc):
* Source/WebCore/accessibility/AXUtilities.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::revealAncestors):
(WebCore::AccessibilityNodeObject::isHiddenUntilFoundContainer const):
(WebCore::AccessibilityNodeObject::revealableText const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::defaultObjectInclusion const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::revealAncestors):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::convertToPropertyFlag):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::objectBecameIgnored):

Canonical link: https://commits.webkit.org/299649@main
